compiler/src/iree/compiler/Codegen/Common/VerifyWorkgroupDistribution.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "iree/compiler/Dialect/HAL/IR/HALTypes.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_VERIFYWORKGROUPDISTRIBUTIONPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+namespace {
+
+/// Pass to verify that all writes to global memory are explicitly mapped to
+/// workgroups. This means that in cases where we use loops (scf.forall) to
+/// manage distribution to workgroups, we require that all ops with write
+/// side effects are contained within a workgroup distributed loop.
+struct VerifyWorkgroupDistributionPass final
+    : impl::VerifyWorkgroupDistributionPassBase<
+          VerifyWorkgroupDistributionPass> {
+
+  void runOnOperation() override {
+    FunctionOpInterface funcOp = getOperation();
+
+    WalkResult hasForall = funcOp.walk([&](Operation *op) {
+      if (auto forallOp = dyn_cast<scf::ForallOp>(op)) {
+        if (forallOpHasMappingType<IREE::Codegen::WorkgroupMappingAttr>(
+                forallOp)) {
+          return WalkResult::interrupt();
+        }
+      }
+      return WalkResult::advance();
+    });
+
+    // Without a workgroup level forall, either this is a single workgroup
+    // dispatch, in which case no verification is needed, or this is already
+    // distributed in which case verification is no longer possible.
+    if (!hasForall.wasInterrupted()) {
+      return;
+    }
+
+    auto globalAddressSpace = IREE::HAL::DescriptorTypeAttr::get(
+        &getContext(), IREE::HAL::DescriptorType::StorageBuffer);
+
+    bool failed = false;
```

Comment:
Can we remove the variable and just check the result of the `funcOp.walk`? E.g.,

```cpp
WalkResult res = funcOp.walk([&](Operation *op) {
// ...
}
if (res.wasInterrupted()) {
  return signalPassFailure();
}
```